lib/benchpark/new_scaling.py

```diff
@@ -0,0 +1,398 @@
+                lambda self: self.weak_scale()
+            ))
+
+        if mode == ScalingMode.Throughput:
+            def throughput_scale(self):
+                raise NotImplementedError(
+                    f"Experiment must implement throughput_scale"
+                )
+            BaseScaling.throughput_scale = throughput_scale
+            scaling_calls.append((
+                lambda self: self.spec.satisfies("scaling=throughput"),
+                lambda self: self.throughput_scale()
+            ))
+
+    def scale(self):
+        for check, action in scaling_calls:
+            if check(self):
+                return action(self)
+        raise RuntimeError("No valid scaling mode matched")
+
+    BaseScaling.scale = scale
+
+    # Helper class
+    class Helper(ExperimentHelper):
+        def get_helper_name_prefix(self):
+            return (
+                f"{self.spec.variants['scaling'][0]}_scaling"
+                if not self.spec.satisfies("scaling=off")
+                else "no_scaling"
+            )
+
+    return type(
+        "ExperimentScaling",
+        (BaseScaling,),
+        {
+            "Helper": Helper,
+        },
+    )
+
+
+def UsesPerProcessDomains(*modes):
+    ScalingType = Scaling(*modes)
+
+    for mode in modes:
+        if mode == ScalingMode.Strong:
+            @requires_experiment_variables("num_procs", "problem_sizes")
+            def strong_scale(self):
+                """
+                Strong scales the problem by increasing the number of processes along each dimension in a
+                round-robin manner, starting with the minimum dimension and decreasing the corresponding
+                per-process size to keep the total problem size constant.
+                Raises an error if scaling down the per-process size does not conserve the global problem size.
+                """
+
+                num_procs = self.expr_vars.num_procs
+                problem_sizes = self.expr_vars.problem_sizes
+
+                if problem_sizes.ndims != num_procs.ndims:
+                    raise BenchparkError(
+                        f"problem_sizes dimensions {problem_sizes.dims} do not match num_procs dimensions {num_procs.dims}"
+                    )
+
+                num_exprs = int(self.spec.variants["scaling-iterations"][0]) - 1
+                scaling_factor = int(self.spec.variants["scaling-factor"][0])
+
+                orig_global_prob_size = num_procs.prod[-1] * problem_sizes.prod[-1]
+               
+                start_dim = num_procs.min_dim
+                ndims = num_procs.ndims
+
+                total_problem_size = [orig_global_prob_size]
+         
+                for itr in range(num_exprs):
+                    dim = (start_dim + itr) % ndims
+
+                    num_procs.scale_dim(dim, lambda v: v * scaling_factor)
+                    problem_sizes.scale_dim(dim, lambda v: v // scaling_factor)
```

Comment:
For initial per-process problem sizes that are not a power of two, this will eventually not cleanly subdivide. You have an error below to catch this, but could this be avoided by taking the ceiling? If not, should this just require up front that the dimension is divisible by `2^x` where `x = num_exprs/ndims`? The outcome is similar, but the error message would be clearer IMO.